File: monitoring/03-carbon-dashboard/carbon_dashboard.ts
/**
 * Carbon metrics dashboard for HarchOS (TypeScript).
 *
 * Builds a comprehensive carbon metrics dashboard combining platform
 * metrics, carbon intensity data, and carbon-aware scheduling results.
 *
 * Usage:
 *   npx ts-node carbon_dashboard.ts
 *   npx ts-node carbon_dashboard.ts --detailed
 *   npx ts-node carbon_dashboard.ts --carbon-budget 5000
 */

import { HarchOSClient } from "harchos";
import type { PlatformMetrics } from "harchos";

// ---------------------------------------------------------------------------
// Simulated data
// ---------------------------------------------------------------------------

interface HubData {
  name: string;
  region: string;
  gpus: number;
  renewablePct: number;
  carbonIntensity: number;
  activeWorkloads: number;
}

const SIMULATED_HUBS: HubData[] = [
  { name: "Harch Ouarzazate", region: "morocco", gpus: 800, renewablePct: 97.2, carbonIntensity: 18, activeWorkloads: 45 },
  { name: "Harch Dakhla", region: "morocco", gpus: 400, renewablePct: 94.8, carbonIntensity: 32, activeWorkloads: 22 },
  { name: "Harch Benguerir", region: "morocco", gpus: 350, renewablePct: 88.5, carbonIntensity: 55, activeWorkloads: 18 },
  { name: "Harch Tanger", region: "morocco", gpus: 200, renewablePct: 82.1, carbonIntensity: 95, activeWorkloads: 8 },
  { name: "Harch Casablanca", region: "morocco", gpus: 48, renewablePct: 45.0, carbonIntensity: 210, activeWorkloads: 3 },
];

const SIMULATED_METRICS: PlatformMetrics = {
  totalHubs: 5,
  totalGpus: 1798,
  availableGpus: 423,
  gpuUtilizationPercent: 76.5,
  totalWorkloads: 1247,
  activeWorkloads: 96,
  totalEnergyKwh: 345678.5,
  avgRenewablePercentage: 81.52,
  avgCarbonIntensity: 47.0,
  avgPue: 1.12,
  totalCo2SavedKg: 12456.3,
};

// ---------------------------------------------------------------------------
// Dashboard logic
// ---------------------------------------------------------------------------

const GREEN_THRESHOLD = 100;
const YELLOW_THRESHOLD = 200;
const INDUSTRY_AVG_INTENSITY = 450;

function carbonIcon(intensity: number): string {
  if (intensity <= GREEN_THRESHOLD) return "🟢";
  if (intensity <= YELLOW_THRESHOLD) return "🟡";
  return "🔴";
}

function computeCarbonSavings(metrics: PlatformMetrics) {
  const co2Industry = (metrics.totalEnergyKwh * INDUSTRY_AVG_INTENSITY) / 1000;
  const co2Harchos = (metrics.totalEnergyKwh * metrics.avgCarbonIntensity) / 1000;
  const co2Saved = co2Industry - co2Harchos;
  const savingsPct = (1 - metrics.avgCarbonIntensity / INDUSTRY_AVG_INTENSITY) * 100;

  return {
    co2IndustryKg: Math.round(co2Industry * 10) / 10,
    co2HarchosKg: Math.round(co2Harchos * 10) / 10,
    co2SavedKg: Math.round(co2Saved * 10) / 10,
    co2SavedTonnes: Math.round((co2Saved / 1000) * 100) / 100,
    savingsPercentage: Math.round(savingsPct * 10) / 10,
  };
}

function detectGreenWindows(hubs: HubData[]) {
  return hubs
    .filter((h) => h.carbonIntensity <= GREEN_THRESHOLD)
    .map((h) => ({
      hub: h.name,
      carbonIntensity: h.carbonIntensity,
      renewablePct: h.renewablePct,
      availableGpus: h.gpus - h.activeWorkloads * 8,
      status: h.carbonIntensity <= 50 ? "optimal" : "green",
    }));
}

function renderDashboard(
  metrics: PlatformMetrics,
  hubs: HubData[],
  detailed: boolean,
): void {
  const savings = computeCarbonSavings(metrics);
  const windows = detectGreenWindows(hubs);

  console.log("\n  ╔════════════════════════════════════════════════════════════════╗");
  console.log("  ║          HarchOS Carbon Metrics Dashboard                     ║");
  console.log("  ╚════════════════════════════════════════════════════════════════╝");

  // Overview
  console.log("\n  🌍 Carbon Overview");
  console.log("  " + "─".repeat(60));
  console.log(`    Average Carbon Intensity: ${metrics.avgCarbonIntensity.toFixed(1).padStart(6)} gCO2/kWh ${carbonIcon(metrics.avgCarbonIntensity)}`);
  console.log(`    Average Renewable Share:  ${metrics.avgRenewablePercentage.toFixed(1).padStart(6)}%`);
  console.log(`    Total Energy Consumed:    ${metrics.totalEnergyKwh.toLocaleString("en", { maximumFractionDigits: 1 }).padStart(10)} kWh`);
  console.log(`    Average PUE:              ${metrics.avgPue.toFixed(2).padStart(6)}`);

  // Carbon Savings
  console.log("\n  💚 Carbon Savings (vs Industry Average)");
  console.log("  " + "─".repeat(60));
  console.log(`    Industry Average:     ${INDUSTRY_AVG_INTENSITY.toString().padStart(6)} gCO2/kWh`);
  console.log(`    HarchOS Average:      ${metrics.avgCarbonIntensity.toFixed(0).padStart(6)} gCO2/kWh`);
  console.log(`    Savings:              ${savings.savingsPercentage.toFixed(1).padStart(5)}%`);
  console.log(`    CO2 Avoided:          ${savings.co2SavedKg.toLocaleString("en", { maximumFractionDigits: 1 }).padStart(10)} kg (${savings.co2SavedTonnes.toFixed(2)} tonnes)`);
  console.log(`    CO2 Saved (Scheduled):${metrics.totalCo2SavedKg.toLocaleString("en", { maximumFractionDigits: 1 }).padStart(10)} kg`);

  // Hub Breakdown
  console.log("\n  🏭 Hub Carbon Breakdown");
  console.log("  " + "─".repeat(60));
  console.log(`    ${"Hub".padEnd(22)} | ${"gCO2/kWh".padStart(8)} | ${"Renewable".padStart(9)} | ${"GPUs".padStart(5)} | Status`);
  console.log(`    ${"─".repeat(22)}-+-${"─".repeat(8)}-+-${"─".repeat(9)}-+-${"─".repeat(5)}-+-${"─".repeat(10)}`);

  const sortedHubs = [...hubs].sort((a, b) => a.carbonIntensity - b.carbonIntensity);
  for (const hub of sortedHubs) {
    const icon = carbonIcon(hub.carbonIntensity);
    const status =
      hub.carbonIntensity <= 50 ? "OPTIMAL" :
      hub.carbonIntensity <= GREEN_THRESHOLD ? "GREEN" :
      hub.carbonIntensity <= YELLOW_THRESHOLD ? "ACCEPTABLE" : "AVOID";
    console.log(
      `    ${hub.name.padEnd(22)} | ${hub.carbonIntensity.toString().padStart(7)}  | ` +
      `${hub.renewablePct.toFixed(1).padStart(8)}% | ${hub.gpus.toString().padStart(5)} | ${icon} ${status}`
    );
  }

  // Green Windows
  if (windows.length) {
    console.log("\n  🌿 Current Green Windows");
    console.log("  " + "─".repeat(60));
    for (const w of windows) {
      console.log(
        `    ${carbonIcon(w.carbonIntensity)} ${w.hub.padEnd(22)} — ` +
        `${w.carbonIntensity} gCO2/kWh | ${w.renewablePct.toFixed(1)}% renewable | ` +
        `~${w.availableGpus} GPUs available | ${w.status.toUpperCase()}`
      );
    }
  }

  // Detailed
  if (detailed) {
    console.log("\n  📈 Detailed Carbon Analysis");
    console.log("  " + "─".repeat(60));

    for (const hub of hubs) {
      const gpuShare = hub.gpus / metrics.totalGpus;
      const hubEnergy = metrics.totalEnergyKwh * gpuShare;
      const hubCo2 = (hubEnergy * hub.carbonIntensity) / 1000;

      console.log(`\n    ${hub.name}:`);
      console.log(`      Estimated Energy:     ${hubEnergy.toLocaleString("en", { maximumFractionDigits: 1 }).padStart(10)} kWh`);
      console.log(`      Estimated CO2:        ${hubCo2.toLocaleString("en", { maximumFractionDigits: 1 }).padStart(10)} kg`);
    }

    // Equivalents
    const treesEquiv = Math.round(savings.co2SavedKg / 22);
    const carsEquiv = (savings.co2SavedKg / 4600).toFixed(1);
    console.log("\n    Carbon Equivalents:");
    console.log(`      CO2 saved = ${treesEquiv.toLocaleString()} trees absorbing for 1 year`);
    console.log(`      CO2 saved = ${carsEquiv} cars off the road for 1 year`);
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const detailed = args.includes("--detailed");
  const carbonBudget = parseFloat(getArg(args, "--carbon-budget", "0")) || undefined;
  const jsonOutput = args.includes("--json");

  console.log("=".repeat(60));
  console.log("HarchOS Carbon Dashboard (TypeScript)");
  console.log("=".repeat(60));

  let metrics: PlatformMetrics = SIMULATED_METRICS;
  let hubs: HubData[] = SIMULATED_HUBS;

  // Try SDK
  const apiKey = process.env.HARCHOS_API_KEY || "";
  if (apiKey) {
    try {
      const client = new HarchOSClient({ apiKey });
      metrics = await client.monitoring.metrics();
      console.log("  [INFO] Using live metrics from HarchOS API");
    } catch (err) {
      console.log(`  [WARN] SDK call failed: ${err} — using simulated data`);
    }
  }

  if (jsonOutput) {
    console.log(JSON.stringify({ metrics, hubs }, null, 2));
  } else {
    renderDashboard(metrics, hubs, detailed);

    // Carbon budget check
    if (carbonBudget) {
      const savings = computeCarbonSavings(metrics);
      const withinBudget = savings.co2HarchosKg <= carbonBudget;

      console.log("\n  💰 Carbon Budget Check");
      console.log("  " + "─".repeat(60));
      console.log(`    Carbon Budget:    ${carbonBudget.toLocaleString().padStart(10)} kg CO2`);
      console.log(`    Current Emissions:${savings.co2HarchosKg.toLocaleString("en", { maximumFractionDigits: 1 }).padStart(10)} kg CO2`);
      console.log(`    Budget Used:      ${((savings.co2HarchosKg / carbonBudget) * 100).toFixed(1).padStart(9)}%`);

      if (withinBudget) {
        console.log("    ✅ Within carbon budget");
      } else {
        console.log("    ❌ Over carbon budget");
        console.log("    💡 Consider scheduling workloads during green windows");
      }
    }

    console.log();
  }
}

function getArg(args: string[], flag: string, defaultValue: string): string {
  const idx = args.indexOf(flag);
  if (idx !== -1 && idx + 1 < args.length) return args[idx + 1];
  return defaultValue;
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: monitoring/03-carbon-dashboard/carbon_dashboard.py
#!/usr/bin/env python3
"""Carbon metrics dashboard for HarchOS.

Builds a comprehensive carbon metrics dashboard using data from
the HarchOS SDK. Combines platform metrics, carbon intensity data,
and carbon-aware scheduling results into a unified view.

Features:
- Platform-wide carbon intensity monitoring
- Per-hub carbon breakdown
- CO2 savings from carbon-aware scheduling
- Green window detection and recommendations
- Carbon budget tracking
- Sovereignty-aware carbon reporting

Usage:
    python carbon_dashboard.py
    python carbon_dashboard.py --detailed
    python carbon_dashboard.py --carbon-budget 5000
    HARCHOS_API_KEY=hsk_... python carbon_dashboard.py
"""

import argparse
import json
import math
import os
from datetime import datetime, timezone
from typing import Dict, List, Optional

try:
    from harchos import HarchOSClient
    HARCHOS_AVAILABLE = True
except ImportError:
    HARCHOS_AVAILABLE = False
    print("[WARN] harchos SDK not installed — using simulated carbon data")


# ---------------------------------------------------------------------------
# Simulated carbon data
# ---------------------------------------------------------------------------

SIMULATED_HUBS = [
    {"name": "Harch Ouarzazate", "region": "morocco", "gpus": 800,
     "renewable_pct": 97.2, "carbon_intensity": 18, "active_workloads": 45},
    {"name": "Harch Dakhla", "region": "morocco", "gpus": 400,
     "renewable_pct": 94.8, "carbon_intensity": 32, "active_workloads": 22},
    {"name": "Harch Benguerir", "region": "morocco", "gpus": 350,
     "renewable_pct": 88.5, "carbon_intensity": 55, "active_workloads": 18},
    {"name": "Harch Tanger", "region": "morocco", "gpus": 200,
     "renewable_pct": 82.1, "carbon_intensity": 95, "active_workloads": 8},
    {"name": "Harch Casablanca", "region": "morocco", "gpus": 48,
     "renewable_pct": 45.0, "carbon_intensity": 210, "active_workloads": 3},
]

SIMULATED_PLATFORM_METRICS = {
    "total_hubs": 5,
    "total_gpus": 1798,
    "available_gpus": 423,
    "gpu_utilization_percent": 76.5,
    "total_workloads": 1247,
    "active_workloads": 96,
    "total_energy_kwh": 345678.5,
    "avg_renewable_percentage": 81.52,
    "avg_carbon_intensity": 47.0,
    "avg_pue": 1.12,
    "total_co2_saved_kg": 12456.3,
}


# ---------------------------------------------------------------------------
# Carbon Dashboard
# ---------------------------------------------------------------------------

class CarbonDashboard:
    """Builds and displays a carbon metrics dashboard for HarchOS."""

    # Thresholds for carbon intensity color coding
    GREEN_THRESHOLD = 100   # gCO2/kWh
    YELLOW_THRESHOLD = 200  # gCO2/kWh

    def __init__(self):
        if HARCHOS_AVAILABLE:
            self.client = HarchOSClient(api_key=os.environ.get("HARCHOS_API_KEY", ""))
        else:
            self.client = None

    def fetch_dashboard_data(self) -> dict:
        """Fetch all data needed for the dashboard."""
        data = {
            "hubs": SIMULATED_HUBS,
            "platform_metrics": SIMULATED_PLATFORM_METRICS,
            "carbon_savings": self._compute_carbon_savings(),
            "green_windows": self._detect_green_windows(),
        }

        if self.client:
            try:
                # Try to get real carbon dashboard data
                carbon_dashboard = self.client.carbon.get_dashboard()
                data["carbon_dashboard"] = {
                    "metrics": {
                        "total_optimizations": getattr(carbon_dashboard, 'metrics', None),
                    }
                }
            except Exception:
                pass

            try:
                metrics = self.client.monitoring.metrics()
                data["platform_metrics"] = {
                    "total_hubs": metrics.total_hubs,
                    "total_gpus": metrics.total_gpus,
                    "available_gpus": metrics.available_gpus,
                    "gpu_utilization_percent": metrics.gpu_utilization_percent,
                    "total_workloads": metrics.total_workloads,
                    "active_workloads": metrics.active_workloads,
                    "total_energy_kwh": metrics.total_energy_kwh,
                    "avg_renewable_percentage": metrics.avg_renewable_percentage,
                    "avg_carbon_intensity": metrics.avg_carbon_intensity,
                    "avg_pue": metrics.avg_pue,
                    "total_co2_saved_kg": metrics.total_co2_saved_kg,
                }
            except Exception:
                pass

        return data

    def _compute_carbon_savings(self) -> dict:
        """Compute carbon savings from using green hubs vs industry average."""
        # Industry average carbon intensity for data centers: ~450 gCO2/kWh
        INDUSTRY_AVG = 450.0
        harchos_avg = SIMULATED_PLATFORM_METRICS["avg_carbon_intensity"]
        total_energy = SIMULATED_PLATFORM_METRICS["total_energy_kwh"]

        # CO2 at industry average
        co2_industry = total_energy * INDUSTRY_AVG / 1000  # kg
        # CO2 at HarchOS average
        co2_harchos = total_energy * harchos_avg / 1000  # kg
        # Savings
        co2_saved = co2_industry - co2_harchos
        savings_pct = (1 - harchos_avg / INDUSTRY_AVG) * 100

        return {
            "co2_industry_kg": round(co2_industry, 1),
            "co2_harchos_kg": round(co2_harchos, 1),
            "co2_saved_kg": round(co2_saved, 1),
            "co2_saved_tonnes": round(co2_saved / 1000, 2),
            "savings_percentage": round(savings_pct, 1),
            "industry_avg_intensity": INDUSTRY_AVG,
            "harchos_avg_intensity": harchos_avg,
        }

    def _detect_green_windows(self) -> List[dict]:
        """Detect current green windows (hubs with low carbon intensity)."""
        windows = []
        for hub in SIMULATED_HUBS:
            if hub["carbon_intensity"] <= self.GREEN_THRESHOLD:
                windows.append({
                    "hub": hub["name"],
                    "carbon_intensity": hub["carbon_intensity"],
                    "renewable_pct": hub["renewable_pct"],
                    "available_gpus": hub["gpus"] - hub["active_workloads"] * 8,
                    "status": "optimal" if hub["carbon_intensity"] <= 50 else "green",
                })
        return windows

    def _carbon_icon(self, intensity: float) -> str:
        """Get color icon for carbon intensity level."""
        if intensity <= self.GREEN_THRESHOLD:
            return "🟢"
        elif intensity <= self.YELLOW_THRESHOLD:
            return "🟡"
        else:
            return "🔴"

    def render_dashboard(self, data: dict, detailed: bool = False):
        """Render the carbon dashboard."""
        print("\n  ╔════════════════════════════════════════════════════════════════╗")
        print("  ║          HarchOS Carbon Metrics Dashboard                     ║")
        print("  ╚════════════════════════════════════════════════════════════════╝")

        metrics = data["platform_metrics"]
        savings = data["carbon_savings"]
        hubs = data["hubs"]
        windows = data["green_windows"]

        # ── Overview ──────────────────────────────────────────────────────
        print(f"\n  🌍 Carbon Overview")
        print(f"  {'─' * 60}")
        print(f"    Average Carbon Intensity: {metrics['avg_carbon_intensity']:>6.1f} gCO2/kWh "
              f"{self._carbon_icon(metrics['avg_carbon_intensity'])}")
        print(f"    Average Renewable Share:  {metrics['avg_renewable_percentage']:>6.1f}%")
        print(f"    Total Energy Consumed:    {metrics['total_energy_kwh']:>10,.1f} kWh")
        print(f"    Average PUE:              {metrics['avg_pue']:>6.2f}")

        # ── Carbon Savings ───────────────────────────────────────────────
        print(f"\n  💚 Carbon Savings (vs Industry Average)")
        print(f"  {'─' * 60}")
        print(f"    Industry Average:     {savings['industry_avg_intensity']:>6.0f} gCO2/kWh")
        print(f"    HarchOS Average:      {savings['harchos_avg_intensity']:>6.0f} gCO2/kWh")
        print(f"    Savings:              {savings['savings_percentage']:>5.1f}%")
        print(f"    CO2 Avoided:          {savings['co2_saved_kg']:>10,.1f} kg "
              f"({savings['co2_saved_tonnes']:>6.2f} tonnes)")
        print(f"    CO2 Saved (Scheduled):{metrics['total_co2_saved_kg']:>10,.1f} kg")

        # ── Hub Breakdown ────────────────────────────────────────────────
        print(f"\n  🏭 Hub Carbon Breakdown")
        print(f"  {'─' * 60}")
        print(f"    {'Hub':22s} | {'gCO2/kWh':>8s} | {'Renewable':>9s} | {'GPUs':>5s} | Status")
        print(f"    {'─' * 22}-+-{'─' * 8}-+-{'─' * 9}-+-{'─' * 5}-+-{'─' * 10}")

        for hub in sorted(hubs, key=lambda h: h["carbon_intensity"]):
            icon = self._carbon_icon(hub["carbon_intensity"])
            status = "OPTIMAL" if hub["carbon_intensity"] <= 50 else (
                "GREEN" if hub["carbon_intensity"] <= self.GREEN_THRESHOLD else (
                    "ACCEPTABLE" if hub["carbon_intensity"] <= self.YELLOW_THRESHOLD else "AVOID"
                )
            )
            print(f"    {hub['name']:22s} | {hub['carbon_intensity']:>7.0f}  | "
                  f"{hub['renewable_pct']:>8.1f}% | {hub['gpus']:>5d} | {icon} {status}")

        # ── Green Windows ────────────────────────────────────────────────
        if windows:
            print(f"\n  🌿 Current Green Windows")
            print(f"  {'─' * 60}")
            for w in windows:
                print(f"    {self._carbon_icon(w['carbon_intensity'])} {w['hub']:22s} — "
                      f"{w['carbon_intensity']:.0f} gCO2/kWh | "
                      f"{w['renewable_pct']:.1f}% renewable | "
                      f"~{w['available_gpus']} GPUs available | {w['status'].upper()}")

        # ── Detailed View ────────────────────────────────────────────────
        if detailed:
            print(f"\n  📈 Detailed Carbon Analysis")
            print(f"  {'─' * 60}")

            # Per-hub energy estimate
            total_energy = metrics["total_energy_kwh"]
            total_gpus = metrics["total_gpus"]
            for hub in hubs:
                gpu_share = hub["gpus"] / total_gpus
                hub_energy = total_energy * gpu_share
                hub_co2 = hub_energy * hub["carbon_intensity"] / 1000  # kg
                hub_renewable_energy = hub_energy * hub["renewable_pct"] / 100

                print(f"\n    {hub['name']}:")
                print(f"      Estimated Energy:     {hub_energy:>10,.1f} kWh")
                print(f"      Estimated CO2:        {hub_co2:>10,.1f} kg")
                print(f"      Renewable Energy:     {hub_renewable_energy:>10,.1f} kWh")

            # Carbon budget equivalent
            print(f"\n    Carbon Equivalents:")
            co2_saved = savings["co2_saved_kg"]
            # Trees absorb ~22 kg CO2/year
            trees_equiv = co2_saved / 22
            # Average car emits ~4600 kg CO2/year
            cars_equiv = co2_saved / 4600
            print(f"      CO2 saved = {trees_equiv:,.0f} trees absorbing for 1 year")
            print(f"      CO2 saved = {cars_equiv:,.1f} cars off the road for 1 year")

    def check_carbon_budget(self, data: dict, budget_kg: float):
        """Check if carbon emissions are within budget."""
        metrics = data["platform_metrics"]
        savings = data["carbon_savings"]

        current_emissions = savings["co2_harchos_kg"]
        within_budget = current_emissions <= budget_kg

        print(f"\n  💰 Carbon Budget Check")
        print(f"  {'─' * 60}")
        print(f"    Carbon Budget:    {budget_kg:>10,.1f} kg CO2")
        print(f"    Current Emissions:{current_emissions:>10,.1f} kg CO2")
        print(f"    Budget Used:      {current_emissions / budget_kg * 100:>9.1f}%")
        print(f"    Remaining:        {max(0, budget_kg - current_emissions):>10,.1f} kg CO2")

        if within_budget:
            print(f"    ✅ Within carbon budget")
        else:
            overage = current_emissions - budget_kg
            print(f"    ❌ Over carbon budget by {overage:,.1f} kg CO2")
            print(f"    💡 Consider scheduling workloads during green windows")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(description="HarchOS Carbon Dashboard")
    parser.add_argument("--detailed", action="store_true",
                        help="Show detailed carbon analysis")
    parser.add_argument("--carbon-budget", type=float, default=None,
                        help="Carbon budget in kg CO2 (optional)")
    parser.add_argument("--json", action="store_true",
                        help="Output raw JSON")
    args = parser.parse_args()

    dashboard = CarbonDashboard()

    print("=" * 60)
    print("HarchOS Carbon Dashboard")
    print("=" * 60)

    data = dashboard.fetch_dashboard_data()

    if args.json:
        print(json.dumps(data, indent=2))
    else:
        dashboard.render_dashboard(data, detailed=args.detailed)

        if args.carbon_budget is not None:
            dashboard.check_carbon_budget(data, args.carbon_budget)

        print()


if __name__ == "__main__":
    main()
